sdk/identity/Azure.Identity/src/ClientCertificateCredential.cs

```diff
@@ -135,5 +183,215 @@ public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext r
+                catch (Exception e) when (!(e is OperationCanceledException))
+                {
+                    throw new CredentialUnavailableException("Could not load certificate file", e);
+                }
+            }
+
+            private async Task<X509Certificate2> LoadCertificateFromPemFileAsync(bool async, string clientCertificatePath, CancellationToken cancellationToken)
+            {
+                if (!(Certificate is null))
+                {
+                    return Certificate;
+                }
+
+                string certficateText;
+
+                try
+                {
+                    if (!async)
+                    {
+                        certficateText = File.ReadAllText(clientCertificatePath);
+                    }
+                    else
+                    {
+                        StringBuilder sb = new StringBuilder();
+                        char[] buf = new char[BufferSize];
+                        int offset = 0;
+
+                        using (StreamReader sr = new StreamReader(clientCertificatePath))
+                        {
+                            while (true)
+                            {
+                                cancellationToken.ThrowIfCancellationRequested();
+                                int read = await sr.ReadAsync(buf, offset, buf.Length).ConfigureAwait(false);
+
+                                for (int i = 0; i < read; i++)
+                                {
+                                    sb.Append(buf[i]);
+                                }
+
+                                if (read == 0)
+                                {
+                                    break;
+                                }
+                            }
+
+                            certficateText = sb.ToString();
+                        }
+                    }
+
+                    Regex certificateRegex = new Regex("(-+BEGIN CERTIFICATE-+)(\n\r?|\r\n?)([A-Za-z0-9+/\n\r]+=*)(\n\r?|\r\n?)(-+END CERTIFICATE-+)", RegexOptions.CultureInvariant);
+                    Regex privateKeyRegex = new Regex("(-+BEGIN PRIVATE KEY-+)(\n\r?|\r\n?)([A-Za-z0-9+/\n\r]+=*)(\n\r?|\r\n?)(-+END PRIVATE KEY-+)", RegexOptions.CultureInvariant);
+
+                    Match certificateMatch = certificateRegex.Match(certficateText);
+                    Match privateKeyMatch = privateKeyRegex.Match(certficateText);
+
+                    if (!certificateMatch.Success)
+                    {
+                        throw new InvalidDataException("Could not find certificate in PEM file");
+                    }
+
+                    if (!privateKeyMatch.Success)
+                    {
+                        throw new InvalidDataException("Could not find private key in PEM file");
+                    }
+
+                    // ImportPkcs8PrivateKey was added in .NET Core 3.0, it is only present on Core.  If we can't find this method, we have a lightweight decoder we can use.
```

Comment:
Should we crosscompile to netcore3.0?

Reply:
I have no intuition here.  It feels like it's not buying us a ton (we'd still have the reflection invoke for `CopyWithPrivateKey`, and I don't know if the complexity of cross compiling is worth it.  But I'm happy to make the change if you feel it's net better).

Reply:
`we'd still have the reflection invoke for CopyWithPrivateKey` why?

Reply:
> we'd still have the reflection invoke for CopyWithPrivateKey why?

Because it's not present in .NET Standard 2.0, but it is present on desktop in .NET 4.7.2 and above.  I suppose we could also cross compile for net472, if we wanted to get around that?

Naively, the late binding approach (at least for `CopyWithPrivateKey`) feels nice to me because things just "light up" if you update your framework, without having to republish your app to a different TFM.

But I'm happy to do whatever ya'll think is best - you've been living the multi-targeting life longer than I have and so if you think it's worthwhile, I can make the changes.

Reply:
We should start by cross targeting tests to include netcoreapp3.1 to at least get some coverage of reflection code path.

One thing I don't like with dynamic light-up is that I don't know if the reflection branch ever gets hit in testing.

Reply:
I like the dynamic light up as well. We've done something similar for the key vault `CryptographyClient` where algorithm support varies by target framework. In this case we left the library only targeted to netstandard2.0 and added the relevant target frameworks to the tests as @pakrym  suggested.

Reply:
> In this case we left the library only targeted to netstandard2.0 and added the relevant target frameworks to the tests as @pakrym suggested.

Thanks - I'll copy the pattern there.